sound/soc/sof/ipc3-topology.c

```diff
@@ -2081,7 +2081,9 @@ static int sof_ipc3_dai_config(struct snd_sof_dev *sdev, struct snd_sof_widget *
 		break;
 	case SOF_DAI_INTEL_ALH:
 		if (data) {
-			config->dai_index = data->dai_index;
+			/* save the dai_index during hw_params and reuse it for hw_free */
+			if (flags & SOF_DAI_CONFIG_FLAGS_HW_PARAMS)
+				config->dai_index = data->dai_index;
```

Comment:
rather unclear to me. We never used the different config stages for ALH, what makes it a requirement now?

Reply:
@plbossart the fact that we never used is what didn't expose the problem. With the new change, we actually send the dai_config ipc properly during hw_params and hw_free. So we need to preserve the dai_index for use during hw_frew

Reply:
sorry, not following @ranj063 

````c
static int sdw_params_stream(struct device *dev,
			     struct sdw_intel_stream_params_data *params_data)
{
	struct snd_soc_dai *d = params_data->dai;
	struct snd_sof_dai_config_data data;
	struct snd_soc_dapm_widget *w;

	w = snd_soc_dai_get_widget(d, params_data->stream);
	data.dai_index = (params_data->link_id << 8) | d->id;
	data.dai_data = params_data->alh_stream_id;

	return hda_ctrl_dai_widget_setup(w, SOF_DAI_CONFIG_FLAGS_NONE, &data);
}

static int sdw_free_stream(struct device *dev,
			   struct sdw_intel_stream_free_data *free_data)
{
	struct snd_soc_dai *d = free_data->dai;
	struct snd_sof_dai_config_data data;
	struct snd_soc_dapm_widget *w;

	w = snd_soc_dai_get_widget(d, free_data->stream);
	data.dai_index = (free_data->link_id << 8) | d->id;

	/* send invalid stream_id */
	data.dai_data = 0xFFFF;

	return hda_ctrl_dai_widget_free(w, SOF_DAI_CONFIG_FLAGS_NONE, &data);
}
````

Where do you see that we will see this SOF_DAI_CONFIG_FLAGS_HW_PARAMS set?

Reply:
@plbossart the dai_config ipc in sdw_free_stream() would never get sent because the dai widget would already have been freed during the FE DAI hw_free in sof_widget_free(). And if you look at that function, we only set the channel not the DAI index because we just don't have that info

Reply:
sorry, I am following even less.

a) if we are already sending the dai_config IPC somewhere else than sdw_free_stream(), then why on earth do we still have this function. we would need a larger change.

b) if we are not sending the dai_config IPC, ever, then the flow is broken.

Which is it?

Reply:
It is a). We should just remove sdw_free_stream()

Reply:
ok, then let's do a real cleanup, it's really confusing to have dead code that remains present.

Reply:
actually wait, where would we send the invalid stream id, if it's not done in sdw_free_stream?

Reply:
Sof_widget_free()

Reply:
it's a nightmare to follow....

we call tplg_ops->dai_config from sdw_hw_params, sof_widget_setup, hda_ipc3_post_trigger, sof_widget_free and sdw_hw_free.

For SoundWire/ALH, this seems completely overkill or redundant. It's quite likely that we were sending the same IPC multiple times for no good reason.

Reply:
I've cleaned up the dai config calls during BE DAI hw_free now. The others will have to stay though.

Reply:
I am still in the dark on the sequences.

For HDaudio, we would still have ipc3_post_trigger and sof_widget_free sending the same dai config (same DMA_CHAN_INVALID, same FLAGS_HW_FREE)

For ALH, we still have a dai config IPC in sdw_hw_params and sof_widget_setup.

Reply:
For AlH, the ipc will only be sent during sof_widget_setup. Sdw_hw_params() will only update the stream ID.

The post_trigger ipc is also needed because we perform the  cleanup during the stop trigger for the HDA DAIs. I'm not sure it is really required. I'll need to experiment to see if I can remove it

Reply:
in that case, why do we set the HW_PARAMS flag in sdw_hw_params()? It's a no-op or rather it'll be done again in sof_widget_setup.

I also don't get what happens in ipc3_post_trigger, the use_count is only decreased in sof_widget_free, so there's no IPC, and I am not sure what cleanup you are referring to.

Reply:
@plbossart for the start/stop test that we have where we do the stop without hw_free, I think we need to leep the post_trigger op to clear the channel ID in the DSP.

As for sdw_hw_params(), its not a no-op. We get the alh_stream_id during BE hw_params and send it during DW hw_params

Reply:
> I also don't get what happens in ipc3_post_trigger, the use_count is only decreased in sof_widget_free, so there's no IPC, and I am not sure what cleanup you are referring to.

@plbossart post_trigger() is called during the BE DAI STOP/SUSPEND trigger. We call hda_link_dma_cleanup() during thse triggers, so we have to clear the channel ID in the DSP as well.

Reply:
I can't follow this sorry. We have 5 stages (hw_params, widget_setup, trigger_stop, widget_free, hw_free) multiplied by two (BE/FE). We don't need this complexity for ALH, and I'll personally forget all the explanations after each long week-end.

Reply:
BTU ALH doesnt have this complexity at all. So i'm not getting your point. For ALH, we have BE hw_params, where we get the stream ID and save it in the config. And send it to the DSP after the DAI widget is setup.

Reply:
maybe what will help make a bit easier to follow is if I moved the dai_config op from being a topology op to being a DAI widget op. But that will take a bit of work. For now, can we please consider mergin this to fix the Brya issue?

Reply:
Commit subject: `fixup! ASoC: SOF: sof-audio: Fix broken early bclk feature for SSP`, emphasis on **SSP**
and the patch modifies `case SOF_DAI_INTEL_ALH:` ?

Reply:
I think the fixup convention is that we use the initial commit title that's being fixed. In this case, we fixed SSP in the initial commit but broke ALH with a side effect. It's still a fixup for the same issue.